src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt

```diff
@@ -0,0 +1,188 @@
+            ))
+        }
+    )
+
+    private fun createSequence(): () -> Long = Instant.now().run {
+        AtomicLong(epochSecond * TimeUnit.SECONDS.toNanos(1) + nano)
+    }::incrementAndGet
+
+    private val firstSequence = createSequence()
+    private val secondSequence = createSequence()
+
+    fun publish(message: RawMessage) {
+        val alias = message.metadata.id.connectionId.sessionAlias
+        val topic = config.aliasToTopic[alias] ?: error("Session alias not found.")
+        val value = message.body.toByteArray()
+
+        val messageID = message.metadata.id.toBuilder()
+            .setBookName(factory.boxConfiguration.bookName)
+            .setDirection(Direction.SECOND)
+            .setSequence(secondSequence())
+            .setTimestamp(Instant.now().toTimestamp())
+            .build()
+
+        messageProcessor.process(
+            RawMessage.newBuilder()
+                .setMetadata(RawMessageMetadata.newBuilder().setId(messageID))
+                .setBody(message.body)
+                .build()
+        )
+
+        val kafkaRecord = ProducerRecord<String, ByteArray>(topic, value)
+        producer.send(kafkaRecord) { _, exception: Throwable? ->
+            when (exception) {
+                null -> {
+                    val msgText = "Message '${message.metadata.id}' sent to Kafka"
+                    LOGGER.info(msgText)
+                    eventSender.onEvent(msgText, "Send message", message)
+                }
+                else -> {
+                    val msgText = "Failed to send message '${message.metadata.id}' to Kafka"
+                    LOGGER.error(msgText, exception)
+                    eventSender.onEvent(msgText, "SendError", message, exception)
+                }
+            }
+        }
+    }
+
+    override fun run() {
+        try {
+            consumer.subscribe(config.topicToAlias.keys)
+            while (!Thread.currentThread().isInterrupted) {
+                val records = consumer.poll(Duration.ofMillis(DURATION_IN_MILLIS))
+
+                if (!records.isEmpty) {
+                    val inactivityPeriod = System.currentTimeMillis() - records.first().timestamp()
+                    if (inactivityPeriod > config.maxInactivityPeriodMillis) {
+                        consumer.seekToEnd(getAllPartitions())
+                        val msgText = "Inactivity period exceeded ($inactivityPeriod ms). Skipping unread messages."
+                        LOGGER.info { msgText }
+                        eventSender.onEvent(msgText, "ConnectivityServiceEvent")
+                    } else {
+                        for (record in records) {
+                            val messageID = factory.newMessageIDBuilder()
+                                .setConnectionId(
+                                    ConnectionID.newBuilder()
+                                        .setSessionAlias(config.topicToAlias[record.topic()])
```

Comment:
Shouldn't we check if the value was not null?

Reply:
It can't be null. We check topicToAlias to have non empty values in Config.kt. And we only subscribe to topics which are keys in topicToAlias.